pxr/base/tf/unicodeUtils.h

```diff
@@ -32,11 +32,60 @@
 #include "pxr/base/tf/diagnostic.h"
 #include "pxr/base/tf/unicodeCharacterClasses.h"
 
+#include <ostream>
 #include <string>
 #include <string_view>
 
 PXR_NAMESPACE_OPEN_SCOPE
 
+/// Wrapper for a code point value that can be encoded as UTF-8
+class TfUtf8CodePoint {
+public:
+    /// Code points that cannot be decoded or outside of the valid range are
+    /// may be replaced with this value.
+    static constexpr uint32_t ReplacementValue = 0xFFFD;
+
+    /// Values higher than this will be replaced with the replacement
+    /// code point.
+    static constexpr uint32_t MaximumValue = 0x10FFFF;
+
+    /// Values in this range (inclusive) cannot be constructed and will be
+    /// replaced by the replacement code point.
+    static constexpr std::pair<uint32_t, uint32_t>
+    SurrogateRange = {0xD800, 0xDFFF};
+
+    /// Construct a code point initialized to the replacement value
+    constexpr TfUtf8CodePoint() = default;
+
+    /// Construct a UTF-8 valued code point, constrained by the maximum value
+    /// and surrogate range.
+    constexpr explicit TfUtf8CodePoint(uint32_t value) :
+        _value(((value <= MaximumValue) &&
+                ((value < SurrogateRange.first) ||
+                 (value > SurrogateRange.second))) ?
+               value : ReplacementValue) {}
+
+    constexpr uint32_t AsUInt32() const { return _value; }
+
+    friend constexpr bool operator==(const TfUtf8CodePoint left,
+                                     const TfUtf8CodePoint right) {
+        return left._value == right._value;
+    }
+    friend constexpr bool operator!=(const TfUtf8CodePoint left,
+                                     const TfUtf8CodePoint right) {
+        return left._value != right._value;
+    }
+
+private:
+    uint32_t _value{ReplacementValue};
+};
+
+TF_API std::ostream& operator<<(std::ostream&, const TfUtf8CodePoint);
+
+/// The replacement code point can be used to signal that a code point could
+/// not be decoded and needed to be replaced.
```

Comment:
It's not a big deal, but I think it would be a little cleaner if this was just
`constexpr TfUtf8CodePoint TfUtf8InvalidCodePoint = TfUtf8CodePoint();`
since the default constructed TfUtf8CodePoint is the invalid code point getting rid of another hardcoded 0xFFFD